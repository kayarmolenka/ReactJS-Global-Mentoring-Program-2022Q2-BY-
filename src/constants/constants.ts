import { IMovieList } from '../models';

export const ADD_MOVIE_TEXT = '+ add movie';
export const valueFilter = ['All', 'Documentary', 'Comedy', 'Horror', 'Crime'];
export const valueSortMovie = ['Select', 'Release date', 'Genre', 'Rating'];

export const INIT_STATE: IMovieList = {
  budget: 0,
  genres: [],
  id: 0,
  overview: '',
  poster_path: '',
  release_date: '',
  revenue: 0,
  runtime: 0,
  tagline: '',
  title: '',
  vote_average: 0,
  vote_count: 0,
};

export const DEFAULT_SRC = require('../assets/images/default_poster.png');
