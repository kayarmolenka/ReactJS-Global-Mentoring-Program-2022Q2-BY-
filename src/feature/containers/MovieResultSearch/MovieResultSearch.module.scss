.movieResultSearch {
  display: flex;
  justify-content: space-between;
  position: relative;
  margin-right: 74px;

  .movieResultSearchBorderLine {
    border-bottom: 2px solid #555555;
    width: 97%;
    position: absolute;
    bottom: 0;
    margin-left: 57px;
  }
}

