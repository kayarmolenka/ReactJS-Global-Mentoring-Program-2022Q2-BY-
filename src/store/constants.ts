import { IApplicationState } from './types';

export const initialState: IApplicationState = {
  movieList: [],
};

export const FETCH_MOVIE_LIST = 'movies/fetchMovieList';
export const FETCH_SORTED_MOVIE_LIST = 'movies/fetchSortedMovieList';
export const FETCH_FILTERED_MOVIE_LIST = 'movies/fetchFilteredMovieList';
export const API_URL = 'http://localhost:4000/';
