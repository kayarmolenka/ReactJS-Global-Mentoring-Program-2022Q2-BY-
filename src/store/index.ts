export * from './store';
export * from './constants';
export * from './thunks';
export * from './types';
export * from './movieListSlice';
export * from './selectors';
