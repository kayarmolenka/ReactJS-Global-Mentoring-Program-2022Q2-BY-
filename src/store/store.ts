import { useDispatch } from 'react-redux';
import { configureStore } from '@reduxjs/toolkit';
import { AppDispatch, RootState } from './index';
import { actionApplicationReducer } from './movieListSlice';

export const reducer = {
  application: actionApplicationReducer,
};

export const createStore = (preloadedState?: Partial<RootState>) =>
  configureStore({ reducer, preloadedState });

export const store = createStore();

export const useAppDispatch = () => useDispatch<AppDispatch>();
