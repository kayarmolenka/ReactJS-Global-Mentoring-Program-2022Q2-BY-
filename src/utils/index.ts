export * from './filterMovies';
export * from './sortMovie';
