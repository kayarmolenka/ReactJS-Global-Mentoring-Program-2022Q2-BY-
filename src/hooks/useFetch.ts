import { IFetchMovieListResponse } from '../store';

export async function useFetch(
  input: RequestInfo,
  init?: RequestInit,
): Promise<IFetchMovieListResponse> {
  const response = await fetch(input, init);
  return response.ok ? response.json() : response;
}
