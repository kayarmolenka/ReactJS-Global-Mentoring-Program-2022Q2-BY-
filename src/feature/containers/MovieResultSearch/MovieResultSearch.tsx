import { MouseEvent, useEffect, useState } from 'react';
import { MovieFilter, MovieSort } from '../../../components';
import {
  fetchFilteredMovieList,
  fetchMovieList,
  fetchSortedMovieList,
} from '../../../store/applications';
import { useAppDispatch } from '../../../store';
import { mapSortsName } from '../../../utils';
import { valueFilter, valueSortMovie } from '../../../constants';

import styles from './MovieResultSearch.module.scss';

export const MovieResultSearch = () => {
  const [chosenTypeSorting, setChosenTypeSorting] = useState(valueSortMovie[0]);
  const [activeGenre, setActiveGenre] = useState(valueFilter[0]);

  const dispatch = useAppDispatch();

  const onHandleGenre = async (e: MouseEvent<HTMLButtonElement>) => {
    setActiveGenre(e.currentTarget.innerHTML);

    if (e.currentTarget.innerHTML !== 'All') {
      await dispatch(
        fetchFilteredMovieList({
          activeGenre: e.currentTarget.innerHTML,
          sortBy: mapSortsName(chosenTypeSorting),
        }),
      );
      return;
    }
    await dispatch(fetchMovieList({ sortBy: mapSortsName(chosenTypeSorting) }));
  };

  useEffect(() => {
    if (chosenTypeSorting && activeGenre !== valueSortMovie[0]) {
      dispatch(
        fetchSortedMovieList({ sortBy: mapSortsName(chosenTypeSorting), activeGenre: activeGenre }),
      );
    }
    if (chosenTypeSorting) {
      dispatch(fetchMovieList({ sortBy: mapSortsName(chosenTypeSorting) }));
    }
  }, [chosenTypeSorting, dispatch]);

  return (
    <div className={styles.movieResultSearch}>
      <MovieFilter activeGenre={activeGenre} onHandleGenre={onHandleGenre} />
      <MovieSort
        chosenTypeSorting={chosenTypeSorting}
        setChosenTypeSorting={setChosenTypeSorting}
      />
      <div className={styles.movieResultSearchBorderLine} />
    </div>
  );
};
