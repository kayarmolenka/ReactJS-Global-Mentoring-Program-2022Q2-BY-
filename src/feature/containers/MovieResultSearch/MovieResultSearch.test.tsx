import { fireEvent, render, screen, waitFor } from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom';
import { Provider } from 'react-redux';
import { MovieResultSearch } from './MovieResultSearch';
import { createStore } from '../../../store';
import { useFetch } from '../../../hooks/useFetch';
import { valueFilter, valueSortMovie } from '../../../constants';
import { mapSortsName } from '../../../utils';

jest.mock('../../../hooks/useFetch', () => ({
  useFetch: jest.fn(),
}));

describe('MovieResultSearch', () => {
  beforeEach(() => {
    (useFetch as jest.Mock).mockReturnValue(Promise.resolve([]));
  });

  afterEach(() => {
    (useFetch as jest.Mock).mockRestore();
  });

  it('should match snapshot', () => {
    const { asFragment } = render(
      <MemoryRouter>
        <Provider store={createStore()}>
          <MovieResultSearch
            activeGenre={valueFilter[0]}
            chosenTypeSorting={valueSortMovie[0]}
            offset={12}
          />
        </Provider>
      </MemoryRouter>,
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it('should call useFetch with filter equal Documentary after click on Documentary button', async () => {
    render(
      <MemoryRouter>
        <Provider store={createStore()}>
          <MovieResultSearch
            activeGenre={valueFilter[1]}
            chosenTypeSorting={mapSortsName(valueSortMovie[0])}
            offset={12}
          />
        </Provider>
      </MemoryRouter>,
    );

    fireEvent.click(screen.getByText('Documentary'));

    await waitFor(() => {
      expect(useFetch as jest.Mock).toHaveBeenCalledWith(
        'http://localhost:4000/movies?sortBy=release_date&sortOrder=desc&offset=1&limit=12',
      );
    });
  });
});
