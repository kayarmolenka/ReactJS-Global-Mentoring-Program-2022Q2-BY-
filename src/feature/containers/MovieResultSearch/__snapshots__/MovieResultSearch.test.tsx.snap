// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MovieResultSearch should match snapshot 1`] = `
<DocumentFragment>
  <div
    class="movieResultSearch"
  >
    <div
      class="movieFilter"
    >
      <button
        class="movieFilterBtn movieFilterBtnChosen"
      >
        All
      </button>
      <button
        class="movieFilterBtn"
      >
        Documentary
      </button>
      <button
        class="movieFilterBtn"
      >
        Comedy
      </button>
      <button
        class="movieFilterBtn"
      >
        Horror
      </button>
      <button
        class="movieFilterBtn"
      >
        Crime
      </button>
    </div>
    <div
      class="movieSort"
    >
      <ul>
        <li
          class="movieSortBy"
        >
          Sort by
        </li>
        <li
          class="movieSortRealiseDate"
        >
          <select
            class="movieSortSelect"
            data-testid="select"
          >
            <option
              value="Release date"
            >
              Release date
            </option>
            <option
              value="Genre"
            >
              Genre
            </option>
            <option
              value="Rating"
            >
              Rating
            </option>
          </select>
          <svg
            aria-hidden="true"
            class="svg-inline--fa fa-caret-down movieSortIconCaret"
            data-icon="caret-down"
            data-prefix="fas"
            focusable="false"
            role="img"
            viewBox="0 0 320 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M310.6 246.6l-127.1 128C176.4 380.9 168.2 384 160 384s-16.38-3.125-22.63-9.375l-127.1-128C.2244 237.5-2.516 223.7 2.438 211.8S19.07 192 32 192h255.1c12.94 0 24.62 7.781 29.58 19.75S319.8 237.5 310.6 246.6z"
              fill="currentColor"
            />
          </svg>
        </li>
      </ul>
    </div>
    <div
      class="movieResultSearchBorderLine"
    />
  </div>
</DocumentFragment>
`;
